geopandas/tools/overlay.py

```diff
@@ -176,3 +176,143 @@ def overlay(df1, df2, how, use_sindex=True):
+        raise ValueError("`how` was \"%s\" but is expected to be in %s" % \
+            (how, allowed_hows))
+
+    if isinstance(df1, GeoSeries) or isinstance(df2, GeoSeries):
+        raise NotImplementedError("overlay currently only implemented for GeoDataFrames")
+
+    if (df1.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df1.index) or 
+        df2.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df2.index)):
+        raise TypeError("overlay only takes GeoDataFrames with (multi)polygon geometries") 
+
+    # Computations
+    df1 = df1.copy()
+    df2 = df2.copy()
+    df1['geometry'] = df1.geometry.buffer(0)
+    df2['geometry'] = df2.geometry.buffer(0)
+    if df1.crs!=df2.crs and reproject:
+        print('Data has different projections.')
+        print('Converted data to projection of first GeoPandas DatFrame')
+        df2.to_crs(crs=df1.crs, inplace=True)
+    if how=='intersection':
+        # Spatial Index to create intersections
+        spatial_index = df2.sindex
+        df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
+        df1['sidx']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
+        pairs = df1['sidx'].to_dict()
+        nei = []
+        for i,j in pairs.items():
+            for k in j:
+                nei.append([i,k])
+        if nei!=[]:
+            pairs = GeoDataFrame(nei, columns=['idx1','idx2'], crs=df1.crs)
+            pairs = pairs.merge(df1, left_on='idx1', right_index=True)
+            pairs = pairs.merge(df2, left_on='idx2', right_index=True, suffixes=['_1','_2'])
+            pairs['Intersection'] = pairs.apply(lambda x: (x['geometry_1'].intersection(x['geometry_2'])).buffer(0), axis=1)
+            pairs = GeoDataFrame(pairs, columns=pairs.columns, crs=df1.crs)
+            cols = pairs.columns.tolist()
+            cols.remove('geometry_1')
+            cols.remove('geometry_2')
+            cols.remove('sidx')
+            cols.remove('bbox')
+            cols.remove('Intersection')
+            dfinter = pairs[cols+['Intersection']].copy()
+            dfinter.rename(columns={'Intersection':'geometry'}, inplace=True)
+            dfinter = GeoDataFrame(dfinter, columns=dfinter.columns, crs=pairs.crs)
+            dfinter = dfinter.loc[dfinter.geometry.is_empty==False]
+            #dfinter.drop(['idx1','idx2'], inplace=True, axis=1)
+            dfinter.reset_index(inplace=True, drop=True)
+            return dfinter
+        else:
+            return GeoDataFrame([], columns=list(set(df1.columns).union(df2.columns)), crs=df1.crs)
+    elif how=='difference':
+        spatial_index = df2.sindex
+        df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
+        df1['sidx']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
+        df1['new_g'] = df1.apply(lambda x: reduce(lambda x, y: x.difference(y).buffer(0), 
+                                 [x.geometry]+list(df2.iloc[x.sidx].geometry)) , axis=1)
```

Comment:
This is a bit difficult to follow. Can you add some comments?

Reply:
Which part? The first two lines are the same as used in another part. Would like to have the same comment all over when repeating a similar operation?